### PR TITLE
feat: native poseidon hashing func

### LINKED ISSUE #304: Poseidon hashing func on native zig
Implement poseidon hashing function on native zig.
Based on: https://github.com/xJonathanLEI/starknet-rs/blob/master/starknet-crypto/src/pedersen_hash.rs

### PR DESCRIPTION
**Pull request by StringNick**
closes #304

### CODE REVIEW

File: src/math/crypto/poseidon/poseidon.zig [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    state[0] = Felt252.add(t, Felt252.mul(Felt252.two(), state[0]));
+    state[1] = Felt252.sub(t, Felt252.mul(Felt252.two(), state[1]));
+    state[2] = Felt252.sub(t, Felt252.mul(Felt252.three(), state[2]));
+}
+
+/// Linear layer for MDS matrix M = ((3,1,1), (1,-1,1), (1,1,2))
+/// Given state vector x, it returns Mx, optimized by precomputing t.
+fn round_comp(state: *[3]Felt252, idx: usize, full: bool) void {
+    if (full) {
+        state[0] = Felt252.add(state[0], COMPRESSED_ROUND_CONSTS[idx]);
```
Comment (tcoratger) [inline src/math/crypto/poseidon/poseidon.zig:23]: maybe an inline for loop here to move from 3 lines to 1 line?

File: src/math/crypto/poseidon/poseidon.zig [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        state[1] = Felt252.add(state[1], COMPRESSED_ROUND_CONSTS[idx + 1]);
+        state[2] = Felt252.add(state[2], COMPRESSED_ROUND_CONSTS[idx + 2]);
+
+        state[0] = Felt252.mul(Felt252.mul(state[0], state[0]), state[0]);
```
Comment (tcoratger) [inline src/math/crypto/poseidon/poseidon.zig:27]: maybe an inline for loop here to move from 3 lines to 1 line? wdt?

File: src/math/crypto/poseidon/poseidon.zig [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+/// Computes the Starknet Poseidon hash of an arbitrary number of [Felt252]s.
+///
+/// Using this function is the same as using [PoseidonHasher].
+pub fn poseidon_hash_many(msgs: []const Felt252) Felt252 {
+    var state = [_]Felt252{Felt252.zero()} ** 3;
+
+    var i: usize = 0;
+    while (i + 1 < msgs.len) {
```
Comment (tcoratger) [inline src/math/crypto/poseidon/poseidon.zig:76]: `while (i + 1 < msgs.len) : (i += 2)` wdt?

File: src/poseidon_consts_gen.zig [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    return config;
+}
+
+// compress_round_constants - compressing round constants
+// caller is owner of result slice and should deinit it
+fn compress_round_constants(allocator: Allocator, config: ConfigJSON, round_constants: [][3]Felt252) ![]Felt252 {
+    var result = std.ArrayList(Felt252).init(allocator);
+
+    for (round_constants[0 .. config.full_rounds / 2]) |rk| {
+        try result.append(rk[0]);
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:58]: inline for 0..3, wdt?

File: src/poseidon_consts_gen.zig [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        try result.append(rk[2]);
+    }
+
+    var idx = config.full_rounds / 2;
+
+    var state = [_]Felt252{Felt252.zero()} ** 3;
+
+    // Add keys for partial rounds
+    for (0..config.partial_rounds) |_| {
+        state[0] = Felt252.add(state[0], round_constants[idx][0]);
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:69]: inline for?

File: src/poseidon_consts_gen.zig [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        // MixLayer
+        state[0] = Felt252.add(st, Felt252.mul(Felt252.two(), state[0]));
+        state[1] = Felt252.sub(st, Felt252.mul(Felt252.two(), state[1]));
+        state[2] = Felt252.sub(st, Felt252.mul(Felt252.two(), state[2]));
+
+        idx += 1;
+    }
+
+    // Add keys for first of the last full rounds
+    state[0] = Felt252.add(state[0], round_constants[idx][0]);
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:89]: inline for 0..3 with:

```zig
state[i] = Felt252.add(state[i], round_constants[idx][i]);
try result.append(state[i]);
```

wdt?

File: src/poseidon_consts_gen.zig [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    state[1] = Felt252.add(state[1], round_constants[idx][1]);
+    state[2] = Felt252.add(state[2], round_constants[idx][2]);
+
+    try result.append(state[0]);
+    try result.append(state[1]);
+    try result.append(state[2]);
+
+    for (round_constants[config.full_rounds / 2 + config.partial_rounds + 1 ..]) |rk| {
+        try result.append(rk[0]);
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:98]: inline for?

File: src/poseidon_consts_gen.zig [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+// parse_config - parsing config from json, allocator should be arena allocator
+fn parse_config(allocator: Allocator, json_spec: []const u8) !ConfigJSON {
+    const config = try std.json.parseFromSliceLeaky(
+        ConfigJSON,
+        allocator,
+        json_spec,
+        .{ .allocate = std.json.AllocWhen.alloc_always },
+    );
+
+    return config;
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:49]: Maybe directly return

```zig
return try std.json.parseFromSliceLeaky(
        ConfigJSON,
        allocator,
        json_spec,
        .{ .allocate = std.json.AllocWhen.alloc_always },
    );
```

File: src/poseidon_consts_gen.zig [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+    for (round_keys) |round_key| {
+        const value = round_key.fe;
+        // writing array felt item
+        try std.fmt.format(array_tpl.writer(), round_constants_block_item, .{ value[0], value[1], value[2], value[3] });
+    }
+
+    var result = std.ArrayList(u8).init(allocator);
+
+    try std.fmt.format(result.writer(), round_constants_block, .{ round_keys.len, try array_tpl.toOwnedSlice(), config.full_rounds, config.partial_rounds });
```
Comment (tcoratger) [inline src/poseidon_consts_gen.zig:35]: As the list is quite long, maybe a comma at the end for fmt?

File: src/math/crypto/poseidon/poseidon.zig [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        x, Felt252.zero(), Felt252.one(),
+    };
+
+    poseidon_permute_comp(&state);
+
+    return state[0];
+}
+
+test "poseidon-hash" {
+    const test_data = [_][3]Felt252{
```
Comment (jobez) [inline src/math/crypto/poseidon/poseidon.zig:104]: can we include the comment from the ported source here? https://github.com/xJonathanLEI/starknet-rs/blob/0857bd6cd3bd34cbb06708f0a185757044171d8d/starknet-crypto/src/poseidon_hash.rs#L150

File: src/math/crypto/poseidon/poseidon.zig [hunk 11]
```diff
@@ -0,0 +1,175 @@
+const std = @import("std");
```
Comment (jobez) [inline src/math/crypto/poseidon/poseidon.zig:1]: can we include a comment with the porting lineage? i.e. we are porting from `stark-rs`, which in turn ported from `pathfinder`

something like: https://github.com/xJonathanLEI/starknet-rs/blob/0857bd6cd3bd34cbb06708f0a185757044171d8d/starknet-crypto/src/poseidon_hash.rs#L2

**Review by jobez — COMMENTED** [review]
two smol comments about comments. great job! the finish line nears!

**Review by jobez — APPROVED** [review]

**Pull request merged**